app/assets/javascripts/admin/addon/components/admin-permalink-form.gjs

```diff
@@ -0,0 +1,257 @@
+import Permalink from "admin/models/permalink";
+
+const TYPE_TO_FIELD_MAP = {
+  topic: "topicId",
+  post: "postId",
+  category: "categoryId",
+  tag: "tagName",
+  user: "userId",
+  external_url: "externalUrl",
+};
+
+export default class AdminFlagsForm extends Component {
+  @service router;
+  @service store;
+  @controller adminPermalinks;
+
+  get isUpdate() {
+    return this.args.permalink;
+  }
+
+  @cached
+  get formData() {
+    if (this.isUpdate) {
+      let permalinkType;
+      let permalinkValue;
+      if (!isEmpty(this.args.permalink.topic_id)) {
+        permalinkType = "topic";
+        permalinkValue = this.args.permalink.topic_id;
+      } else if (!isEmpty(this.args.permalink.post_id)) {
+        permalinkType = "post";
+        permalinkValue = this.args.permalink.post_id;
+      } else if (!isEmpty(this.args.permalink.category_id)) {
+        permalinkType = "category";
+        permalinkValue = this.args.permalink.category_id;
+      } else if (!isEmpty(this.args.permalink.tag_name)) {
+        permalinkType = "tag";
+        permalinkValue = this.args.permalink.tag_name;
+      } else if (!isEmpty(this.args.permalink.external_url)) {
+        permalinkType = "external_url";
+        permalinkValue = this.args.permalink.external_url;
+      } else if (!isEmpty(this.args.permalink.user_id)) {
+        permalinkType = "user";
+        permalinkValue = this.args.permalink.user_id;
+      }
+
+      return {
+        url: this.args.permalink.url,
+        [TYPE_TO_FIELD_MAP[permalinkType]]: permalinkValue,
+        permalinkType,
+      };
+    } else {
+      return {
+        permalinkType: "topic",
+      };
+    }
+  }
+
+  get header() {
+    return this.isUpdate
+      ? "admin.permalink.form.edit_header"
+      : "admin.permalink.form.add_header";
+  }
+
+  @action
+  save(data) {
+    const createOrUpdate = this.isUpdate ? this.update : this.create;
+    createOrUpdate(data);
+  }
+
+  @bind
+  async create(data) {
+    try {
+      const result = await this.store.createRecord("permalink").save({
+        url: data.url,
+        permalink_type: data.permalinkType,
+        permalink_type_value: this.valueForPermalinkType(data),
+      });
+      this.adminPermalinks.model.unshiftObject(
+        Permalink.create(result.payload)
+      );
+      this.router.transitionTo("adminPermalinks");
+    } catch (error) {
+      popupAjaxError(error);
+    }
+  }
+
+  @bind
+  async update(data) {
+    try {
+      const result = await this.store.update(
+        "permalink",
+        this.args.permalink.id,
+        {
+          url: data.url,
+          permalink_type: data.permalinkType,
+          permalink_type_value: this.valueForPermalinkType(data),
+        }
+      );
+      const index = this.adminPermalinks.model.findIndex(
+        (permalink) => permalink.id === this.args.permalink.id
+      );
+      this.adminPermalinks.model[index] = Permalink.create(result.payload);
```

Comment:
is this synchronous?

Reply:
Yes, it is just creating an object from args returned by Ajax call.  It is required because there are computed properties that are used to display the permalink row:

https://github.com/discourse/discourse/blob/main/app/assets/javascripts/admin/addon/models/permalink.js#L33

https://github.com/discourse/discourse/blob/main/app/assets/javascripts/admin/addon/templates/permalinks.hbs#L58